experiment/nursery/diskcache/cache.go

```diff
@@ -0,0 +1,136 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// cache implements disk backed cache storage for use in nursery
+package diskcache
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// ReadHandler should be implemeted by cache users for use with Cache.Get
+type ReadHandler func(exists bool, contents io.ReadSeeker) error
+
+// Cache implements disk backed cache storage
+type Cache struct {
+	diskRoot string
+}
+
+// NewCache returns a new Cache given the root directory that should be used
+// on disk for cache storage
+func NewCache(diskRoot string) *Cache {
+	return &Cache{
+		diskRoot,
+	}
+}
+
+func (c *Cache) keyToPath(key string) string {
+	return filepath.Join(c.diskRoot, key)
+}
+
+// DiskRoot returns the root directory containing all on-disk cache entries
+func (c *Cache) DiskRoot() string {
+	return c.diskRoot
+}
+
+// file path helper
+func exists(path string) bool {
+	_, err := os.Stat(path)
+	return !os.IsNotExist(err)
+}
+
+// file path helper
+func ensureDir(dir string) {
+	if exists(dir) {
+		return
+	}
+	log.WithError(os.MkdirAll(dir, os.FileMode(0744))).Infof("MkDirAll(%s)", dir)
```

Comment:
this function probably should return error if `os.MkdirAll` fails? Or do we still want to proceed?

Reply:
We log the error, we don't want the cache to go down for errors, and we will catch the failure to create the file if the directories don't exist. If they exist we don't really care?

Reply:
sounds like you can combine this with temp file creation as `ensureTempFile` or something?
/bikeshed

Reply:
returned error and logged in caller